### ISSUE TITLE
Add a prompt function (variable) for template files

### ISSUE THREAD

**Issue by setedivento** [RealRaven2000/SmartTemplate4#81]
**THIS IS A FEATURE REQUEST.**

When using template files to compose new messages, it would be very useful to have a – say –
&nbsp;&nbsp;&nbsp;&nbsp;`%prompt("someVariableID", "somePromptTitle")%`
function to allow you to store a user input variable in memory, and then use it in the mail body, or other sections, with – say – another function like that:  
&nbsp;&nbsp;&nbsp;&nbsp;`%evaluate("someVariableID")%`.

For example, let's suppose you have this template:
```
%prompt("userText", "Please input your text here")%
%header.set(to, "tryme@xxx.com")%
%header.set(subject, "Issue number: %evaluate('userText')%")%
Hello, i'm a template.
Here is some user input text: %evaluate("userText")%.
Here is the same text again: %evaluate("userText")%
```
SmartTemplate should:
- replace the prompted text in any section it's been declared, be it the subject, body, from, cc or to.
- replace the prompted text any times it's been declared in those sections (e.g. 2 times in the body section).

Of course:
1. this line `%header.set(subject, "Issue number: %evaluate('userText')%")%` would require nested %%-variables evaluation
2. an _evaluate_ function would pose security issues, but if limited to the defined variables scope, and checked against code injection, I think it could work very well.

Thank you very much!

**Comment by RealRaven2000** [comment]
I like it! 

I will look into this once we have done the Tb78 migration. Make sure to remind me again afterwards in case I forget it,

however, `I don't think I will be able to do a nested syntax` like this

`this line %header.set(subject, "Issue title: %evaluate('userText')%")%`

as SmartTemplates is mostly based on regular expression parsing / replacement. But we can of course work on some alternative syntax, avoiding the `%something%` pattern on the "inside".

**Comment by setedivento** [comment]
> I like it!

Cool! :)
&nbsp;
> I will look into this once we have done the Tb78 migration. Make sure to remind me again afterwards in case I forget it,

Yep, indeed.

&nbsp;
> however, `I don't think I will be able to do a nested syntax` like this
> 
> `this line %header.set(subject, "Issue title: %evaluate('userText')%")%`
> 
> as SmartTemplates is mostly based on regular expression parsing / replacement. But we can of course work on some alternative syntax, avoiding the `%something%` pattern on the "inside".

Sure, mine was just an introductory thought line but yes, this feature requires a little bit of engineering.

The first solution that comes to my mind is something like:
```
%header.set(subject, "Issue title: ${userText}")%
%header.replace(subject, "${userText}", userText) %
```
That way, no evaluate function is needed.

**Comment by RealRaven2000** [comment]
More brainstorming - If we have the luxury of declaring variables within the template I think the following syntax could work:

```
%form.set(variable,"value")%
%form.prompt(variable,"default")%
%form.get(variable,"default")%
```

"default" would be an optional parameter. For `%from.prompt%` it could pre-populate the textbox. For `%form.get%` it could fill in a placeholder text if the variable is not set (or an empty string?). I think all variables should be stored globally in the composer window, so that they also work if you have the form initialization in an external template (e.g. by injecting `%file(customerFormVariables.html)%` )

The syntax could also later be extended to initialize a variable by finding a regular expression pattern from the replied / forwarded email, such as:

```
%form.match(variable,body,"regular expression")%
%form.match(variable,subject,"regular expression")%
```

**Comment by setedivento** [comment]
Yep yep, your option seems more consistent with the existing API.
Looking forward to the implementation!
